deep_qa/common/params.py

```diff
@@ -1,52 +1,184 @@
+        choices.  `name` is an optional description for where a configuration error happened, if there
+        is one.
+
+        Note that this _pops_ the key from params, modifying the dictionary, consistent with how
+        parameters are processed in this codebase.
+        """
+
+        value = self.pop(key)
+        if value not in choices:
+            raise ConfigurationError(_get_choice_error_message(value, choices, self.history))
+        return value
+
+    def pop_choice_with_default(self,
+                                key: str,
+                                choices: List[Any],
+                                default: Any=None):
+        """
+        Like get_choice, but with a default value.  If `default` is None, we use the first item in
+        `choices` as the default.
+        """
+        try:
+            return self.pop_choice(key, choices)
+        except KeyError:
+            if default is None:
+                default = choices[0]
+            if default not in choices:
+                raise ConfigurationError(_get_choice_error_message(default, choices, self.history))
+
+            return self.__check_is_dict(key, default)
 
+    def as_dict(self):
+        """
+        Sometimes we need to just represent the parameters as a dict, for instance when we pass
+        them to a Keras layer(so that they can be serialised).
+        """
+        def log_recursively(parameters, history):
+            for key, value in parameters.items():
+                if isinstance(value, dict):
+                    new_local_history = value if history == "" else history + "." + key
+                    log_recursively(value, new_local_history)
+                else:
+                    logger.param(self.history + self.history + " : " + key + " : " + str(value))
 
-def get_choice_with_default(params: Dict[str, Any],
+        logger.info("Converting Params object to dict; logging of default "
+                    "values will not occur when dictionary parameters are "
+                    "used subsequently.")
+        logger.info("CURRENTLY DEFINED PARAMETERS: ")
+        log_recursively(self.params, self.history)
+        return self.params
+
+    def __check_is_dict(self, new_history, value):
+        if isinstance(value, dict):
+            new_history = new_history if self.history == "" else self.history + "." + new_history
+            return Params(value, new_history)
+        else:
+            return value
+
+    def __getitem__(self, key):
+        return self.__check_is_dict(key, self.params[key])
+
+    def __setitem__(self, key, value):
+        self.params[key] = value
+
+    def __delitem__(self, key):
+        del self.params[key]
+
+    def __iter__(self):
+        return iter(self.params)
+
+    def __len__(self):
+        return len(self.params)
+
+
+def _get_choice_error_message(value: Any, choices: List[Any], name: str=None) -> str:
```

Comment:
This is actually a good example of why we shouldn't do this: using double underscores in function names and then using them within methods of a class doesn't work, because when you use a double underscore _anywhere_ in a class, all that happens is that `__function_name` gets replaced with `_classname_function_name`. However, if `__function_name` is not a method on the class, it won't be defined.